modules/generative-openai/clients/openai.go

```diff
@@ -71,22 +72,46 @@ func (v *openai) GenerateAllResults(ctx context.Context, textProperties []map[st
 func (v *openai) Generate(ctx context.Context, cfg moduletools.ClassConfig, prompt string) (*ent.GenerateResult, error) {
 	settings := config.NewClassSettings(cfg)
 
-	body, err := json.Marshal(generateInput{
-		Prompt:           prompt,
-		Model:            settings.Model(),
-		MaxTokens:        settings.MaxTokens(),
-		Temperature:      settings.Temperature(),
-		FrequencyPenalty: settings.FrequencyPenalty(),
-		PresencePenalty:  settings.PresencePenalty(),
-		TopP:             settings.TopP(),
-	})
-	if err != nil {
-		return nil, errors.Wrapf(err, "marshal body")
+	var body []byte
+	var err error
+	var oaiUrl string
+
+	if settings.IsLegacy() {
+		oaiUrl, err = url.JoinPath(v.host, "/v1/completions")
+		if err != nil {
+			return nil, errors.Wrap(err, "join OpenAI API host and path")
+		}
+		body, err = json.Marshal(generateInput{
+			Prompt:           prompt,
+			Model:            settings.Model(),
+			MaxTokens:        settings.MaxTokens(),
+			Temperature:      settings.Temperature(),
+			FrequencyPenalty: settings.FrequencyPenalty(),
+			PresencePenalty:  settings.PresencePenalty(),
+			TopP:             settings.TopP(),
+		})
+	} else {
+		oaiUrl, err = url.JoinPath(v.host, v.path)
+		if err != nil {
+			return nil, errors.Wrap(err, "join OpenAI API host and path")
+		}
+		tokens := determineTokens(settings.GetMaxTokensForModel(settings.Model()), settings.MaxTokens(), prompt)
+		body, err = json.Marshal(generateInput{
+			Messages: []message{{
+				Role:    "user",
+				Content: prompt,
+			}},
+			Model:            settings.Model(),
+			MaxTokens:        float64(tokens),
+			Temperature:      settings.Temperature(),
+			FrequencyPenalty: settings.FrequencyPenalty(),
+			PresencePenalty:  settings.PresencePenalty(),
+			TopP:             settings.TopP(),
+		})
```

Comment:
inside the if statements you can generate the payload (legacy or current one) and invoke the `json.Marhal` method only once in line 112:
```
body, err = json.Marshal(generateInput)
```